src/main/java/frc/robot/subsystems/swerve/SwerveSubsystem.java

```diff
@@ -408,37 +411,31 @@ public void periodic() {
 
     updateOdometry();
     updateVision();
-
-    Logger.recordOutput("Odometry/Fused Pose", estimator.getEstimatedPosition());
-    Logger.recordOutput(
-        "Odometry/Fused to Odo Deviation", estimator.getEstimatedPosition().minus(pose));
   }
 
   private void updateOdometry() {
-    for (int i = 0; i < modules.length; i++) {
-      Logger.recordOutput(
-          "Swerve/Updates Since Last " + i, modules[i].getOdometryTimestamps().length);
-    }
-    double[] sampleTimestamps =
-        modules[0].getOdometryTimestamps(); // All signals are sampled together
-    int sampleCount =
-        Arrays.stream(modules).mapToInt(m -> m.getOdometryTimestamps().length).min().getAsInt();
-    // if the gyro isnt connected, dont worry about it
-    if (gyroInputs.connected) {
-      sampleCount = Math.min(sampleCount, gyroInputs.odometryTimestamps.length);
-    }
-    for (int deltaIndex = 0; deltaIndex < sampleCount; deltaIndex++) {
+    Logger.recordOutput("Swerve/Updates Since Last", odoThreadInputs.sampledStates.size());
+    var sampleStates = odoThreadInputs.sampledStates;
+    for (var sample : sampleStates) {
       // Read wheel deltas from each module
       SwerveModulePosition[] modulePositions = new SwerveModulePosition[4];
       SwerveModulePosition[] moduleDeltas =
           new SwerveModulePosition[4]; // change in positions since the last update
       boolean hasNullModulePosition = false;
+      // Technically we could have not 4 modules worth of data here but im not dealing w that
       for (int moduleIndex = 0; moduleIndex < 4; moduleIndex++) {
-        modulePositions[moduleIndex] =
-            modules[moduleIndex]
-                .getOdometryPositions()[deltaIndex]; // gets positions from the thread, NOT inputs
-        // we can't do any odo updates if we are not getting module data
-        if (modulePositions[moduleIndex] == null) {
+        try {
+          modulePositions[moduleIndex] =
+              new SwerveModulePosition(
+                  sample.values().get(new SignalID(SignalType.DRIVE, moduleIndex)),
+                  Rotation2d.fromRotations(
+                      sample
+                          .values()
+                          .get(
+                              new SignalID(
+                                  SignalType.STEER,
+                                  moduleIndex)))); // gets positions from the thread, NOT inputs
+        } catch (NullPointerException e) {
```

Comment:
This catch is overly broad and I'd suggest you instead handle nulls where they may occur. The problem with just catching the NPE is that you could have NPEs which are the result of a mistake of the code vs expected null values being passed and now you've papered over them like they're the same thing. There be dragons here.